driver/src/driver/scheduler.rs

```diff
@@ -0,0 +1,247 @@
+use super::stablex_driver::StableXDriver;
+use crate::metrics::StableXMetrics;
+use crate::orderbook::StableXOrderBookReading;
+use anyhow::{anyhow, Context, Result};
+use log::error;
+use log::info;
+use std::collections::HashSet;
+use std::thread;
+use std::time::{Duration, Instant, SystemTime, SystemTimeError};
+
+const BATCH_DURATION: Duration = Duration::from_secs(300);
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+struct BatchId(pub u64);
+
+impl BatchId {
+    pub fn current(now: SystemTime) -> std::result::Result<Self, SystemTimeError> {
+        let time_since_epoch = now.duration_since(SystemTime::UNIX_EPOCH)?;
+        Ok(Self(time_since_epoch.as_secs() / BATCH_DURATION.as_secs()))
+    }
+
+    pub fn currently_being_solved(now: SystemTime) -> std::result::Result<Self, SystemTimeError> {
+        Self::current(now).map(|batch_id| batch_id.prev())
+    }
+
+    pub fn start_time(self) -> SystemTime {
+        SystemTime::UNIX_EPOCH + Duration::from_secs(self.0 * BATCH_DURATION.as_secs())
+    }
+
+    pub fn solve_start_time(self) -> SystemTime {
+        self.start_time() + BATCH_DURATION
+    }
+
+    pub fn next(self) -> BatchId {
+        self.0.checked_add(1).map(BatchId).unwrap()
+    }
+
+    pub fn prev(self) -> BatchId {
+        self.0.checked_sub(1).map(BatchId).unwrap()
+    }
+}
+
+pub struct Scheduler<'a> {
+    driver: &'a mut dyn StableXDriver,
+    orderbook_reader: &'a dyn StableXOrderBookReading,
+    metrics: &'a StableXMetrics,
+    batch_wait_time: Duration,
+    max_batch_elapsed_time: Duration,
```

Comment:
Does it make sense to combine these into a AucitonTimingConfiguration struct (it would allow explicit documentation and not require to pass to identical types in this method which are easy to invert)?

Also a nit about naming here, can we call it `auction_processing_start_time` and `solution_submission_time`. I find batch_wait_time too generic in that it could mean anything.

Reply:
Made this a struct and changed names to target_start_solve_time and latest_solve_attempt_time.